### PR TITLE
Similarity Augmented Graph Matching

### LINKED ISSUE #347: similarity augmented graph matching

### PR DESCRIPTION
**Pull request by asaadeldin11**
Closes  #347 
Allows `GraphMatch` to accept a similarity matrix and perform similarity augmented graph matching.

See [notebook](https://nbviewer.jupyter.org/github/asaadeldin11/nbview/blob/master/sim_aug_test.ipynb) for proof of effectiveness

### CODE REVIEW

**Comment by asaadeldin11** [comment]
@dwaynepryce @bdpedigo @alyakin314 @Nyecarr @j1c  i think this is ready for review if anyone has time to give it a look. Thanks!

File: graspologic/match/gmp.py
Comment (Nyecarr) [inline graspologic/match/gmp.py:111]: I think we need to update the docs to reflect the new default.

File: tests/test_match.py
Comment (Nyecarr) [inline tests/test_match.py:9]: We should take a look at restructuring the tests folder -- this is in the root directory but would probably be better if our tests structure resembles the structure of the modules in graspologic. graspologic/match/qap -> tests/match/test_match vs tests/test_match. Shouldn't hold up this pr of course.

**Review by Nyecarr — APPROVED** [review]
I'll leave the math verification on the JHU side but looks good to me other than the documentation comment :)

File: graspologic/match/gmp.py
Comment (bdpedigo) [inline graspologic/match/gmp.py:163]: done or no?

File: graspologic/match/gmp.py
Comment (bdpedigo) [inline graspologic/match/gmp.py:185]: can you provide more info here? also, does the "scale" of this matrix (i.e. frob norm) matter? id like to know that in the docs

File: graspologic/match/gmp.py
Comment (bdpedigo) [inline graspologic/match/gmp.py:249]: reminder to copy doc changes from above to here

File: tests/test_match.py
Comment (bdpedigo) [inline tests/test_match.py:163]: maybe the kinda thing that would be better in tutorial - but does vanilla gm not succeed here in this setting? j.w. cause i havent seen proof that the stuff in this pr works (but I trust you, dont need to go through that whole process). 

also if there's no tutorial for this PR,  can you at least make an issue to add one in the future?

Comment (asaadeldin11) [reply tests/test_match.py:163]: The best proof of effectiveness is found [here](https://github.com/asaadeldin11/nbview/blob/master/simaug_pr_poc.ipynb), where I recreate a figure that Youngser had made.

This figure is also the basis for the test. For rho = 1, vanilla gm would here, which is a good point, so I've changed it to rho = 0.9, and thresholded it such that vanilla gm likely wouldn't work (accuracy > 0.7).

I can add a tutorial showing how with the 0.9-ER graph's, GM+SS gives higher accuracy than vanilla gm?

Comment (bdpedigo) [reply tests/test_match.py:163]: sure. but maybe the tutorial in a separate PR (can make an issue for this) cause i dont want to hold this PR up

Comment (asaadeldin11) [reply tests/test_match.py:163]: sounds good

File: graspologic/match/gmp.py
Comment (asaadeldin11) [reply graspologic/match/gmp.py:185]: size should be the number of nonseeded nodes, so I'll add that to tests and docs. Don't think the norm needs to be restricted.

**Comment by bdpedigo** [comment]
@asaadeldin11 can you resolve all open comments so we can merge? would like to use this for something

File: graspologic/match/gmp.py
Comment (bdpedigo) [reply graspologic/match/gmp.py]: im not saying it needs to be restricted. but i want to know if i multiply S by 2 does that affect the optimization. I think yes?

Comment (asaadeldin11) [reply graspologic/match/gmp.py]: ok i get what you mean now. so if you want both terms to be weighted the same norm(A)*norm(B) should be equal to norm(S)?

Comment (bdpedigo) [reply graspologic/match/gmp.py]: i guess im not saying we need to do anything about this, just that we should point this out to the user

Comment (asaadeldin11) [reply graspologic/match/gmp.py]: something like: "Note, the scale of `S` may effect the weight placed on the term :math:`\\text{trace}(S^T P)` relative to :math:`\\text{trace}(A^T PBP^T)` during the optimization process."

Comment (bdpedigo) [reply graspologic/match/gmp.py]: works for me

File: graspologic/match/qap.py
Comment (asaadeldin11) [inline graspologic/match/qap.py:174]: what's more readable, this or `np.trace(S[:, perm])` ?

Comment (asaadeldin11) [reply graspologic/match/qap.py:174]: other than this I think it's ready to merge

Comment (bdpedigo) [reply graspologic/match/qap.py:174]: trace is more readable but what you have is fine and possibly faster? totally up to you imo

Comment (asaadeldin11) [reply graspologic/match/qap.py:174]: Yeah I made it like this for speed reasons

Comment (asaadeldin11) [reply graspologic/match/qap.py:174]: I'll just leave it as is

**Comment by bdpedigo** [comment]
looks like black failing

**Comment by asaadeldin11** [comment]
> looks like black failing

should be resolved now

File: graspologic/match/gmp.py
Comment (bdpedigo) [inline graspologic/match/gmp.py:163]: random state is usually passed to constructor in sklearn as far as I know. what are your thoughts @asaadeldin11?

File: graspologic/match/gmp.py
Comment (bdpedigo) [inline graspologic/match/gmp.py:183]: what does this mean when A and B have different shape? Can't it be N_A by N_B?

File: graspologic/match/gmp.py
Comment (bdpedigo) [inline graspologic/match/gmp.py:185]:
```suggestion
            Note: the scale of `S` affects the weight placed on the term
```

File: graspologic/match/gmp.py
Comment (bdpedigo) [inline graspologic/match/gmp.py:224]: idk what we'll decide about rng but not that it is not passed here

File: tests/test_match.py
Comment (bdpedigo) [inline tests/test_match.py:163]: is there a sensible test where this is smaller? Been trying to streamline the tests a lot

Comment (asaadeldin11) [reply tests/test_match.py:163]: i can lower n to 75 maybe? GM on 150 nodes is pretty quick though

Comment (asaadeldin11) [reply tests/test_match.py:163]: @bdpedigo thoughts?

Comment (bdpedigo) [reply tests/test_match.py:163]: sure 75 seems fine?

File: graspologic/match/gmp.py
Comment (asaadeldin11) [reply graspologic/match/gmp.py:183]: yeah you're right, should be n_a x n_b

File: graspologic/match/gmp.py
Comment (asaadeldin11) [reply graspologic/match/gmp.py:163]: sure, I'll change it

Comment (asaadeldin11) [reply graspologic/match/gmp.py:163]: should i also change it to `random_state` for consistency with sklearn @bdpedigo ?

Comment (bdpedigo) [reply graspologic/match/gmp.py:163]: yes

**Comment by bdpedigo** [comment]
@asaadeldin11 can you resolve conflicts and resolve any comments above that you addressed

File: graspologic/match/gmp.py
Comment (bdpedigo) [inline graspologic/match/gmp.py:194]:
```suggestion
            A similarity matrix. Should be shape (n_A , n_B), the number of nodes in
```

File: graspologic/match/gmp.py
Comment (bdpedigo) [inline graspologic/match/gmp.py:270]:
```suggestion
            A similarity matrix. Should be shape (n_A , n_B), the number of nodes in
```

File: graspologic/match/qap.py
Comment (bdpedigo) [inline graspologic/match/qap.py:282]:
```suggestion
        A similarity matrix. Should be same shape as ``A`` and ``B``.   
```

**Review by bdpedigo — APPROVED** [review]

**Comment by bdpedigo** [comment]
@asaadeldin11 tests failing

**Comment by asaadeldin11** [comment]
failing tests should be resolved

**Pull request merged**
